[profiler][refactor] Avoid using legacy event in profiler

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61931
* __->__ #61721

Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

Commit:
[profiler][refactor] Avoid using legacy event in profiler

Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

guotuofeng commented on torch/autograd/profiler.py:
```diff
@@ -1148,9 +1149,9 @@ def parse_kineto_results(result):
             for mem_record in mem_records:
                 if (mem_record[0].start_us() >= kineto_event.start_us() and
                         mem_record[0].start_us() <= abs_end_us):
```
this piece of code has performance penalty as we tested. 

https://github.com/pytorch/pytorch/issues/58983 
https://github.com/pytorch/pytorch/issues/61041
https://github.com/pytorch/kineto/issues/308 are complaining this performance issue when handling memory events.

https://github.com/pytorch/pytorch/pull/60432 improve the performance of memory paring

ilia-cher replied:
I'll rebase after your fix lands

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

kimishpatel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -529,6 +518,10 @@ int64_t KinetoEvent::cudaElapsedUs() const {
 }
 
 c10::DeviceType KinetoEvent::deviceType() const {
+  if (device_type_ >= 0) {
```
Why make this `int8_t` and defaulted `-1`? Can you add invalid device type? Or this is tied, 1:1, to device types on aten side so we have to follow that?

kimishpatel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -529,6 +518,10 @@ int64_t KinetoEvent::cudaElapsedUs() const {
 
 c10::DeviceType KinetoEvent::deviceType() const {
+  if (device_type_ >= 0) {
+    return (c10::DeviceType)device_type_;
```
Ok ignore my earlier comment. I see you cast to `c10::DeviceType`

kimishpatel commented on torch/csrc/autograd/profiler_kineto.cpp:
```diff
@@ -182,16 +180,11 @@ struct KinetoThreadLocalState : public ProfilerThreadLocalState {
+  uint64_t start_time_;
   std::vector<KinetoEvent> kineto_events_;
   std::unique_ptr<libkineto::CpuTraceBuffer> cpu_trace;
-  std::vector<libkineto::GenericTraceActivity> memory_events_;
```
Does this mean that support for memory profiling is removed?

ilia-cher replied:
no it doesn't, we pass the memory events in kineto_events_ (as we should've)

kimishpatel commented on torch/autograd/profiler.py:
```diff
@@ -1226,20 +1229,21 @@ def parse_kineto_results(result):
-                scope=mem_record[0].scope(),
-                cpu_memory_usage=mem_record[0].cpu_memory_usage(),
-                cuda_memory_usage=mem_record[0].cuda_memory_usage(),
+                scope=0,  # function
```
Can you add a comment to point to `aten/src/ATen/record_function.h` where RecordScope is and 0 = Function?

kimishpatel commented on torch/autograd/profiler.py:
```diff
@@ -1134,29 +1135,29 @@ def rewrite_name(name, with_wildcard=False):
-            mem_records.append([record, False])
-    assert start_record is not None, "Invalid profiler output, __start_profile is missing"
+    trace_start_us = result.trace_start_us()
+    mem_records = [[evt, False] for evt in result.events() if evt.name() == "[memory]"]
```
What does 'False' mean in `[evt, False]`?

ilia-cher replied:
it's a flag that is used later in the code to filter out "top-level" memory events (events that did not happen in the context of another op), that are outputted separately

kimishpatel reviewed (approved):
Looks good to me. Left some minor comments.

Commit:
Update on "[profiler][refactor] Avoid using legacy event in profiler"


Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

[ghstack-poisoned]

ilia-cher commented:
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).
